File: kernel/src/proc/exec.rs
use crate::mem;
use crate::mem::virt::PageDirFlags;
use crate::drivers::ata;
use crate::fs;
use super::elf;
use log::info;

/// This is sloppily implemented, in order to quickly get to the goal of actually exec'ing a binary
pub fn exec(fname: &str) {
    let ata1 = ata::AtaBlockDevice::new(ata::DriveSelect::Master);
    let ustar = fs::ustar::UStarFs::new(ata1);
    let file = ustar.open(fname).unwrap();

    info!("Exec {}... [size: {}]", file.name, file.size);
    let buf = file.read().unwrap();

    unsafe {
        let elf_header = *(buf.as_ptr() as *const elf::Elf32);
        info!("num prog headers {} entsize {}", elf_header.e_phnum, elf_header.e_phentsize);

        // Create a virtual page that contains the start address. This is extremely hacky.
        let phys_addr = mem::physical::alloc_contiguous_4mb().unwrap();
        let virt_addr = elf_header.e_entry & 0xFFC0000;
        mem::virt::VirtualManager::alloc_4mb(virt_addr, phys_addr, PageDirFlags::WRITE | PageDirFlags::PRESENT | PageDirFlags::_4M_PAGE);

        for phnum in 0..elf_header.e_phnum {
            let off = elf_header.e_phoff as isize + (phnum * elf_header.e_phentsize) as isize;
            let prog_header = *(buf.as_ptr().offset(off) as *const elf::Elf32ProgramHeader);
            if prog_header.p_type == 1 {
                info!("off {:#x} file size {:#x}; mem size {:#x}",prog_header.p_offset, prog_header.p_filesz, prog_header.p_memsz);

                // Ensure that the ELF doesn't want to be loaded outside of the virtual memory we set up
                assert!(prog_header.p_vaddr as u32 >= virt_addr && prog_header.p_vaddr < virt_addr + 4*1024*1024);
                info!("loading at {:#x}", prog_header.p_vaddr as u32);
                compiler_builtins::mem::memcpy(prog_header.p_vaddr as *mut u8, buf.as_ptr().offset(prog_header.p_offset as isize), prog_header.p_filesz as usize);
            }
        }

        info!("Jumping into {} @ {:#x}...", fname, elf_header.e_entry);
        // Jump into the program we just loaded
        asm!("jmp {}", in(reg) elf_header.e_entry as u32);
    }
}

File: src/kernel/proc/exec.c
#include <stdint.h>
#include <kernel/proc/exec.h>
#include <kernel/proc/elf.h>
#include <kernel/stdio.h>
#include <kernel/mem/physical.h>
#include <kernel/mem/virtual.h>
#include <kernel/string.h>

void exec(uint8_t *buf) {
    Elf32 *elf_header = (Elf32*) buf;
    printf("num prog headers %d entsize %d\n", elf_header->e_phnum, elf_header->e_phentsize);

    AllocResult alloc_res = alloc_contiguous_4mb();
    if (alloc_res.status != ALLOC_SUCCESS) {
        printf("error allocating memory\n");
        return;
    }
    uint32_t virt_addr = elf_header->e_entry & 0xFFC0000;
    alloc_4mb(virt_addr, alloc_res.addr, PAGEDIR_WRITE | PAGEDIR_PRESENT | PAGEDIR__4M_PAGE);

    for (int phnum = 0; phnum < elf_header->e_phnum; phnum++) {
        int off = elf_header->e_phoff + (phnum * elf_header->e_phentsize);
        Elf32ProgramHeader *prog_header = (Elf32ProgramHeader*)(buf + off);
        if (prog_header->p_type == 1) {
            printf("off %x file size %x; mem size %x\n",prog_header->p_offset, prog_header->p_filesz, prog_header->p_memsz);
            printf("loading at %x\n", prog_header->p_vaddr);
            memcpy((void*)prog_header->p_vaddr, buf + prog_header->p_offset, prog_header->p_filesz);
        }
    }

    printf("Jumping to %x", elf_header ->e_entry);
    asm volatile (
        "jmp %0"
        :: "r" (elf_header->e_entry)
    );
}